File: MyScript/edk/java/examples/text/CreateSubsetKnowledge.java
package text;

import java.io.File;
import java.io.IOException;

import certificates.MyCertificate;

import com.myscript.engine.Engine;
import com.myscript.text.SubsetKnowledge;
import common.CFG;

/**
 * The <code>CreateSubsetKnowledge</code> example code shows you the basic steps
 * to create a <code>SubsetKnowledge</code> resource that constrains the
 * characters a handwriting recognizer is allowed to recognize.
 */
public class CreateSubsetKnowledge
{
  /**
   * Main entry point for the <code>CreateSubsetKnowledge</code> example
   * program.
   * 
   * @param args command line arguments, ignored.
   */
  public static final void main(String[] args) throws IOException
  {
    System.out.println("1) creating a new Engine with your user certificate");
    final Engine engine = Engine.create(MyCertificate.getBytes());
    System.out.println(" . Engine created successfully");
    System.out.println();
    
    System.out.println("2) creating a new SubsetKnowledge");
    final SubsetKnowledge sk = SubsetKnowledge.create(engine);
    System.out.println(" . SubsetKnowledge created successfully");
    System.out.println();

    System.out.println("3) disabling all characters in the SubsetKnowledge");
    sk.disableAllCharacters();
    System.out.println(" . all characters disabled successfully");
    System.out.println();
    
    System.out.println("4) enabling \"abcdefghijklmnopqrstuvwxyz\" characters");
    sk.enableCharacters("abcdefghijklmnopqrstuvwxyz");
    System.out.println(" . characters enabled successfully");
    System.out.println();
    
    System.out.println("5) storing the SubsetKnowledge as a resource file:");
    sk.store(CFG.OUTPUT_DIR + "/CreateSubsetKnowledge-output.res");
    System.out.println(" . " + new File(CFG.OUTPUT_DIR + "/CreateSubsetKnowledge-output.res").getCanonicalPath() + " stored successfully");
    System.out.println();

    // do not forget the 1st rule: you create it, you dispose of it!
    System.out.println("6) cleaning-up: disposing of objects and disposing of the Engine");
    sk.dispose();
    engine.dispose();
    System.out.println(" . clean-up done successfully");
    System.out.println();
    
    System.out.println("end of example:");
    System.out.println(" you just learnt how to create, edit, store and dispose of a SubsetKnowledge");
  }
  
} // CreateSubsetKnowledge

File: MyScript/edk/c/examples/text/CreateSubsetKnowledge.c
/**
 * The `CreateSubsetKnowledge` example code shows you the basic steps to create
 * a `SubsetKnowledge` resource that constrains enables you to specify the
 * characters a handwriting recognizer is allowed to recognize.
 */


// -- configuration ------------------------------------------------------------

// common configuration parameters
#include "../common/config.h"

// include the MyScript Engine API header(s)
#include <c/MyScriptEngine.h>
#include <c/MyScriptText.h>

// include the certificate
#include "../certificates/MyCertificate.h"


// -- example body -------------------------------------------------------------

#include <stdio.h>

// global engine (for the sake of readability)
voEngine engine = NULL;

// common utility functions
#include "../common/common.h"

// main function
int main()
{
  voSubsetKnowledge sk = NULL;
  voString string = voString_INITIALIZER("abcdefghijklmnopqrstuvwxyz");

  // -- library initialization -------------------------------------------------
  if (!loadMyScriptEngine(MYSCRIPT_ENGINE_LIBRARY_DIR, NULL))
    return (fprintf(stderr, "failed to load the MyScriptEngine library\n"), EXIT_FAILURE);

  // -- example code -----------------------------------------------------------
  printf("1) creating a new Engine initialized with your certificate\n");
  engine = voCreateEngine(VO_MSE_VER, &myCertificate, NULL);
  verify(engine != NULL);
  printf(" . Engine created successfully\n");
  printf("\n");
    
  printf("2) creating a new SubsetKnowledge\n");
  sk = voCreateSubsetKnowledge(engine);
  verify(sk != NULL);
  printf(" . SubsetKnowledge created successfully\n");
  printf("\n");

  printf("3) disabling all characters in the SubsetKnowledge\n");
  verify(VO_IFACE(engine, ISubsetKnowledge)->disableAllCharacters(engine, sk));
  printf(" . all characters disabled successfully\n");
  printf("\n");

  printf("4) enabling \"abcdefghijklmnopqrstuvwxyz\" characters\n");
  verify(VO_IFACE(engine, ISubsetKnowledge)->enableCharacters(engine, sk, NULL, &string));
  printf(" . characters enabled successfully\n");
  printf("\n");

  printf("5) storing the SubsetKnowledge as a resource file:\n");
  storeEngineObject(sk, OUTPUT_DIR "/CreateSubsetKnowledge-output.res");
  printf(" . SubsetKnowldedge stored successfully to: ");
  printPath(stdout, OUTPUT_DIR "/CreateSubsetKnowledge-output.res");
  printf("\n");
  printf("\n");

  printf("6) cleaning-up: releasing all created objects and destroying the Engine\n");
  verify(voReleaseObject(engine, sk));
  voDestroyEngine(engine);
  printf(" . clean-up done successfully\n");
  printf("\n");

  printf("end of example:\n");
  printf(" you just learnt how to create, edit, store release a SubsetKnowledge\n");

  // -- finalization -----------------------------------------------------------
  unloadMyScriptEngine();

  return EXIT_SUCCESS;
}
